### PR TITLE
Add clipping plane in basic viewer

### PR DESCRIPTION
**Pull request by gdamiand**
_Please use the following template to help us managing pull requests._

## Summary of Changes

Add a mechanism in the basic viewer to enable/disable a clipping plane.
Small feature  added by Jing Yang during gsoc 2020.

## Release Management

* Affected package(s): GraphicsView
* Feature/Small Feature (if any): <s>[Small feature page on CGAL wiki](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Clipping_Plane_in_Basic_Viewer)</s> (Edit by @lrineau, 2021/03/08: this pull-request is actually not a small feature.)

### CODE REVIEW

File: GraphicsView/include/CGAL/Basic_shaders.h [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+void main(void)
+{
+  gl_FragColor = vec4(0.0, 0.0, 0.0, 1.0);
+}
+)DELIM";
+*/
+
+}
+
+#endif // CGAL_BASIC_SHADERS_H
```
Comment (maxGimeno) [inline GraphicsView/include/CGAL/Basic_shaders.h:277]: I think the shaders should be in glsl version 150, like in #5336. Mac OS machines have some troubles with the version 120.

**Review by maxGimeno — CHANGES_REQUESTED** [review]
Why not have used a manipulatedFrame for the plane? Maybe it is only because I am used to the Polyhedron demo, but I really don't like the manipulation here. I can understand the translation limited to the direction of the plane, as it is a plane, but at least the rotation is far to brutal to my taste. Actually,  all demos that use planes work the same, not just the Polyhedron_demo, so I really think it should be the same in the Basic Viewer

**Comment by gdamiand** [comment]
> Why not have used a manipulatedFrame for the plane? Maybe it is only because I am used to the Polyhedron demo, but I really don't like the manipulation here. I can understand the translation limited to the direction of the plane, as it is a plane, but at least the rotation is far to brutal to my taste. Actually, all demos that use planes work the same, not just the Polyhedron_demo, so I really think it should be the same in the Basic Viewer

Thanks for the suggestion. It would be better. Just a pity to suggest this now the code is finished... I need to work on this to update...

File: GraphicsView/include/CGAL/Basic_shaders.h (hunk 1 above)
Comment (gdamiand) [reply GraphicsView/include/CGAL/Basic_shaders.h:277]: Do you think it is enough to replace #version 120 by 150 ?

Comment (maxGimeno) [reply GraphicsView/include/CGAL/Basic_shaders.h:277]: Sadly no, you need to replace the `varying` and `attribute` specifiers too.
`attribute` becomes `in`, it is easy to replace, but varrying becomes `out` in vertex shaders and `in` in fragment shaders. 
You must also rpelace the pre-built `gl_FragColor` variable by a user-defind `out` variable in the fragment shaders. Just ask if you want me to take care of that, I did it quite a lot. 
You can also check the diff in #5336, it should be quite clear.

**Comment by lrineau** [comment]
> Just a pity to suggest this now the code is finished...

One cannot blame Maxime. He was not mentoring the student, and as far as I know he was never consulted during the GSoC internship.

File: GraphicsView/include/CGAL/Basic_shaders.h (hunk 1 above)
Comment (gdamiand) [reply GraphicsView/include/CGAL/Basic_shaders.h:283]: Done in https://github.com/CGAL/cgal/pull/5406/commits/a7a3931f0ac67a730a7a063b53010066b5061526

But there are some errors, faces are now all black.
@maxGimeno Any idea where I am wrong please ?

**Comment by gdamiand** [comment]
> > Just a pity to suggest this now the code is finished...
> 
> One cannot blame Maxime. He was not mentoring the student, and as far as I know he was never consulted during the GSoC internship.

I don't blame Maxime; I just regret the late remark. 

For now I suggest to keep the current plane behavior, and integrate the small feature like that.  Indeed, this small feature adds transparency which is nice. Modifications to use manipulatedFrame can be done later.

**Comment by maxGimeno** [comment]
> > > Just a pity to suggest this now the code is finished...
> > 
> > 
> > One cannot blame Maxime. He was not mentoring the student, and as far as I know he was never consulted during the GSoC internship.
> 
> I don't blame Maxime; I just regret the late remark.
> 
> For now I suggest to keep the current plane behavior, and integrate the small feature like that. Indeed, this small feature adds transparency which is nice. Modifications to use manipulatedFrame can be done later.

As it is usable as is, I'm ok with this.

**Comment by afabri** [comment]
The "can be done later"  means probably never.  So we have two different behaviors which is a pity.

**Comment by gdamiand** [comment]
> The "can be done later" means probably never. So we have two different behaviors which is a pity.

Agree; but waiting the modifications means probably never. So we don't have transparency and drop the gsoc which is a pity too. And for my point of view it is better to have two different behaviors that no transparency: question of choice...

**Comment by gdamiand** [comment]
> Those 4 `out` should be `in`

Thanks  @maxGimeno ; I think it is ok now.

**Comment by lrineau** [comment]
> > The "can be done later" means probably never. So we have two different behaviors which is a pity.
> 
> Agree; but waiting the modifications means probably never. So we don't have transparency and drop the gsoc which is a pity too. And for my point of view it is better to have two different behaviors that no transparency: question of choice...

I suggest you collaborate with Maxime, who knows a lot about those topic. Maxime can implement the required modifications.

**Comment by gdamiand** [comment]
@maxGimeno I started here https://github.com/CGAL/cgal/pull/5406/commits/a09a4684f4aa726ccd0f2777654e5af12163cb6a, adding some todo.
Could you help me/give me some advice please ?

**Comment by maxGimeno** [comment]
@gdamiand OK, I'll implement the frame usage, and tell you when I'm done.

**Comment by gdamiand** [comment]
> @gdamiand OK, I'll implement the frame usage, and tell you when I'm done.

Thanks @maxGimeno !!

**Comment by maxGimeno** [comment]
@gdamiand I'm done. You can check you're ok with my changes.  I didn't use drawGrid as you sugggested because it is specific for the viewer's feature that draws the Z plane grid with X and Y axis.

**Comment by gdamiand** [comment]
> @gdamiand I'm done. You can check you're ok with my changes. I didn't use drawGrid as you sugggested because it is specific for the viewer's feature that draws the Z plane grid with X and Y axis.

I tested it, it works perfectly. Thanks again @maxGimeno  !!

**Comment by maxGimeno** [comment]
Has it been reviewed ? When will it be accepted ?

**Comment by maxGimeno** [comment]
Also there are conflicts with master

**Comment by gdamiand** [comment]
> Has it been reviewed ? When will it be accepted ?

As said by Laurent on the small feature page, there is no api and no documentation, thus nothing to accept. You reviewed the code.

**Comment by maxGimeno** [comment]
Ok, I got confused by the flags but Laurent just updated them, let's test this.

**Comment by maxGimeno** [comment]
Errors: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.3-Ic-68/Linear_cell_complex_Examples/TestReport_Blake_Windows_MSVCPreview-Release-64bits.gz

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.3-Ic-74.shtml

**Comment by gdamiand** [comment]
@lrineau Why did you added "not yet approved" flag ? Do you want me to pre-approve this SF ?
You wrote in the wiki page "This small feature cannot really be objected or approved"; I was thinking you agree that there is no need for preapproval ?

**Comment by lrineau** [comment]
@gdamiand The process is rather simple.
  1. For the moment, this small feature not yet approved, and that is why I have added https://github.com/CGAL/cgal/labels/not%20yet%20approved. Violet labels are for the release management: they signal that a pull-request cannot be merged in `master`. 
  2. Once the small feature is pre-approved, it will gain https://github.com/CGAL/cgal/labels/Pre-approved but will retain https://github.com/CGAL/cgal/labels/not%20yet%20approved, that prevents the merge.
  3. Once the small feature is accepted, https://github.com/CGAL/cgal/labels/not%20yet%20approved will be removed, and https://github.com/CGAL/cgal/labels/accepted%20small%20feature will be added.

**Comment by gdamiand** [comment]
> @gdamiand The process is rather simple.
> 
> 1. For the moment, this small feature not yet approved, and that is why I have added https://github.com/CGAL/cgal/labels/not%20yet%20approved. Violet labels are for the release management: they signal that a pull-request cannot be merged in `master`.
> 2. Once the small feature is pre-approved, it will gain https://github.com/CGAL/cgal/labels/Pre-approved but will retain https://github.com/CGAL/cgal/labels/not%20yet%20approved, that prevents the merge.
> 3. Once the small feature is accepted, https://github.com/CGAL/cgal/labels/not%20yet%20approved will be removed, and https://github.com/CGAL/cgal/labels/accepted%20small%20feature will be added.

Thanks @lrineau for your answer. Process simple, but a little bit confusing when you wrote "This small feature cannot really be objected or approved"... I just pre-approved the SF.

**Comment by lrineau** [comment]
> > @gdamiand The process is rather simple.
> > 
> > 1. For the moment, this small feature not yet approved, and that is why I have added [not yet approved (label)](https://github.com/CGAL/cgal/labels/not%20yet%20approved). Violet labels are for the release management: they signal that a pull-request cannot be merged in `master`.
> > 2. Once the small feature is pre-approved, it will gain [Pre-approved (label)](https://github.com/CGAL/cgal/labels/Pre-approved) but will retain [not yet approved (label)](https://github.com/CGAL/cgal/labels/not%20yet%20approved), that prevents the merge.
> > 3. Once the small feature is accepted, [not yet approved (label)](https://github.com/CGAL/cgal/labels/not%20yet%20approved) will be removed, and [accepted small feature (label)](https://github.com/CGAL/cgal/labels/accepted%20small%20feature) will be added.
> 
> Thanks @lrineau for your answer. Process simple, but a little bit confusing when you wrote "This small feature cannot really be objected or approved"... I just pre-approved the SF.

You are right. The confusion comes from that: that pull request is not a small feature. I forgot about that. Let's clean the situation.

**Comment by gdamiand** [comment]
Thanks @lrineau !

**Comment by lrineau** [comment]
Sorry for what happened. Me and Maxime have to deal with hundreds of pull-requests, and last week I had to merge about 30 of them. We cannot remember all the details for every pull-request. That is why we have that system of labels, that help classify the pull-request more quickly. In that system, a small feature that is not approved has the label https://github.com/CGAL/cgal/labels/not%20yet%20approved and cannot be merged. I forgot that this was a peculiar case. I have merged the pull-request as it is.

**Pull request merged**